iphone/Classes/TiLogServer.m

```diff
@@ -0,0 +1,347 @@
+#include <sys/socket.h>
+#include <netinet/in.h>
+
+/**
+ * These constants are available to allow xcodebuild to override them.
+ */
+#ifndef TI_LOG_SERVER_PORT
+# define TI_LOG_SERVER_PORT 10571
+#endif
+
+#ifndef TI_LOG_SERVER_BACKLOG
+# define TI_LOG_SERVER_BACKLOG 4 // max simulataneous connections
+#endif
+
+#ifndef TI_LOG_SERVER_QUEUE_SIZE
+# define TI_LOG_SERVER_QUEUE_SIZE 100
+#endif
+
+/**
+ * All log server state is global mainly because there should only be one log
+ * server.
+ */
+static dispatch_queue_t logDispatchQueue = nil;
+static dispatch_source_t logDispatchSource = nil;
+static int logServerSocket = -1;
+static NSMutableArray* connections = nil;
+static NSMutableArray* logQueue = nil;
+
+/**
+ * Helper function to force logging with NSLog() since we override it. Used only
+ * for reporting log server errors.
+ */
+static void forcedNSLog(NSString* str, ...)
+{
+	va_list args;
+	va_start(args, str);
+	NSString* msg = [[[NSString alloc] initWithFormat:str arguments:args] autorelease];
+#pragma push
+#undef NSLog
+	NSLog(@"%@", msg);
+#pragma pop
+}
+
+/**
+ * Tries to set the socket as non-blocking. Because we're using GCD, this
+ * doesn't appear to be necessary, so if it fails, we can continue.
+ */
+static void trySetNonBlocking(int socket, const char* type) {
+	int flags = fcntl(socket, F_GETFL, 0);
+	int result = fcntl(socket, F_SETFL, flags | O_NONBLOCK);
+	if (result == -1) {
+		forcedNSLog(@"[WARN] Couldn\'t set %s socket to be non-blocking (%d)", type, result);
+	}
+}
+
+/**
+ * A simple queue for storing log messages when there are no active connections.
+ */
+@interface NSMutableArray (TiLogMessageQueue)
+- (id)pop;
+- (void)push:(id)obj;
+@end
+
+@implementation NSMutableArray (TiLogMessageQueue)
+-(id)pop
+{
+	if ([self count]) {
+		id head = [self objectAtIndex:0];
+		if (head != nil) {
+			[self removeObjectAtIndex:0];
+		}
+	}
+}
+
+-(void)push:(id)message
+{
+	// ensure we don't put too many messages in the queue
+	if ([self count] + 1 > TI_LOG_SERVER_QUEUE_SIZE) {
+		[self pop];
+	}
+	[self addObject:message];
+}
+@end
+
+/**
+ * Encapsulates connection state.
+ */
+@interface Connection : NSObject {
+	int socket;
+	dispatch_source_t readSource;
+}
+-(id)initWithSocket:(int)_socket;
+-(void)send:(dispatch_data_t*)buffer;
+@end
+
+@implementation Connection
+
+/**
+ * Initializes a connection with a raw socket file descriptor. This function
+ * will create a read stream on the socket to detect when the client has
+ * disconnected.
+ */
+-(id)initWithSocket:(int)_socket
+{
+	socket = _socket;
+	readSource = nil;
+
+	// disable SIGPIPE signal
+	setsockopt(socket, SOL_SOCKET, SO_NOSIGPIPE, &(int){ 1 }, sizeof(int));
+
+	trySetNonBlocking(socket, "connection\'s");
+
+	// the only way to know if the remote peer has disconnected is to read from
+	// from the socket
+	readSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_READ,
+										socket,
+										0, // mask not used
+										logDispatchQueue);
```

Comment:
That's the default Xcode behavior on my machine. If I hit return after the comma, that's where it indents to. I couldn't find any good examples in the code of indenting long lines because there are a handful of places that favor everything on a single line over readability.

I would be open to formatting as:

```
readSource = dispatch_source_create(
    DISPATCH_SOURCE_TYPE_READ,
    socket,
    0, // mask not used
    logDispatchQueue
);
```

Thoughts?